src/SecureHeaders.php

```diff
@@ -881,22 +875,114 @@ public function hpkproSubdomains($mode = true)
 
     public function done()
     {
-        $this->importHeaders();
-        $this->applyAutomaticHeaders();
+        // TODO: Move this elsewhere!
+        $directive = $this->canInjectStrictDynamic();
+        if ( ! is_int($directive)) {
+            $this->csp($directive, 'strict-dynamic');
+        } else if ($directive !== -1) {
+            $this->addError(
+                "<b>Strict-Mode</b> is enabled, but <b>'strict-dynamic'</b>
+                    could not be added to the Content-Security-Policy because
+                    no hash or nonce was used.",
+                E_USER_WARNING
+            );
+        }
+
+        $this->apply($this->httpAdapter);
+
+        // TODO: Take care of these!
+        #$this->reportMissingHeaders();
+        #$this->validateHeaders();
+        #$this->reportErrors();
+    }
```

Comment:
Indeed 😉 

The header validation will hinge on finishing up the header parse.
Quite impressed with the attribute parsing/writing from what I've read so far though – you've managed to condense most of the functionality into something a lot more readable than the conditional regular expressions I was using to parse each header into attributes 😉 

I guess the only thing missing is decomposing the CSP (and CSPRO) header into directive name/source list string pairs (this to say that it'll need special casing for the parse). I'll try finish up reading through all the changes, and then I can probably help out with some commits if you haven't already added it in by then 😜